cranelift-codegen/src/regalloc/branch_splitting.rs

```diff
@@ -0,0 +1,228 @@
+        cur: EncCursor::new(func, isa),
+        domtree,
+        topo,
+        cfg,
+    };
+    ctx.run()
+}
+
+struct Context<'a> {
+    // True if new blocks were inserted.
+    new_blocks: bool,
+
+    // Current instruction as well as reference to function and ISA.
+    cur: EncCursor<'a>,
+
+    // References to contextual data structures we need.
+    domtree: &'a mut DominatorTree,
+    topo: &'a mut TopoOrder,
+    cfg: &'a mut ControlFlowGraph,
+}
+
+impl<'a> Context<'a> {
+    fn run(&mut self) {
+        // Any ebb order will do.
+        self.topo.reset(self.cur.func.layout.ebbs());
+        while let Some(ebb) = self.topo.next(&self.cur.func.layout, self.domtree) {
+            // Branches can only be at the last or second to last position in an
+            // extended basic block.
+            self.cur.goto_last_inst(ebb);
+            let terminator_inst = self
+                .cur
+                .current_inst()
+                .expect("missing terminator instruction");
+            match self.cur.prev_inst() {
+                Some(inst) => {
+                    let opcode = self.cur.func.dfg[inst].opcode();
+                    if opcode.is_branch() {
+                        self.visit_conditional_branch(inst, opcode);
+                    } else {
+                        continue;
+                    }
+                }
+                // If there is only a terminator instruction, then there is no
+                // need to create a new block.
+                None => continue,
+            };
+            self.cur.goto_inst(terminator_inst);
+            self.visit_terminator_branch(terminator_inst);
+        }
+
+        // If blocks were added the cfg and domtree are inconsistent and must be recomputed.
+        if self.new_blocks {
+            self.cfg.compute(&self.cur.func);
+            self.domtree.compute(&self.cur.func, self.cfg);
+        }
+    }
+
+    fn visit_conditional_branch(&mut self, inst: Inst, opcode: Opcode) {
+        let target = match self.cur.func.dfg[inst] {
+            InstructionData::Branch { destination, .. }
+            | InstructionData::BranchIcmp { destination, .. }
+            | InstructionData::BranchInt { destination, .. }
+            | InstructionData::BranchFloat { destination, .. } => destination,
+            _ => panic!("Unexpected instruction in classify_branch"),
+        };
+
+        // If there are any parameters, split the edge.
+        if self.should_split_edge(inst, target) {
+            // Remember the branch arguments.
+            let jump_args: Vec<Value> = self
+                .cur
+                .func
+                .dfg
+                .inst_variable_args(inst)
+                .iter()
+                .map(|x| *x)
+                .collect();
+
+            // Create the block the branch will jump to.
+            let new_ebb = self.make_empty_ebb();
+
+            // Remove the arguments from the branch and make it jump to the new block.
```

Comment:
@bnjbvr or @sunfishcode  any idea if there is a way to only patch the jump target without changing the rest of the isntruction. My understanding is that we have to call the builder once more and provide all the inputs. Maybe we can have an instruction builder which takes the default from one instruction, and allow us to patch it after, before building it?

Reply:
Can you get a mutable reference on `self.cur.func.dfg[inst]`, match on it that it's one of the expected formats (the different formats ought to be grouped in a single case, since we're only interested in the destination ebb, if it has the same name for every format), and then just modify the destination ebb?

Reply:
I think I found a generic way to do that for any branch, based on the assumption that the variable arguments would always be the Ebb parameters. We can already replace the `inst_data.branchdestination_mut()`, but for the arguments I think we have to rely on `inst_data.take_value_list()` and `inst_data.put_value_list(…, …)`.